### PR TITLE
Improve EntitlementInfo's init

### LINKED ISSUE #774: Refactor EntitlementInfo to use Codable
We pull from dictionaries: 
```swift
entitlementData: [String: Any],
productData: [String: Any],
```

We should migrate this over to Codable

### PR DESCRIPTION
**Pull request by Juanpe**
Resolves #774 

This PR is a proposal to improve the decoding of the Entitlement data coming from the backend using `Codable`. We can't use `Codable` without a large refactor but as @taquitos said [here](https://github.com/RevenueCat/purchases-ios/issues/774#issuecomment-905743066), and I agree, just decoding the `entitlementData` and `productData` using `Codable` would be a great improvement.

TBH, I'm not proud of this code but I think it's a good first improvement 🙂

Thank you for contributing to Purchases. Before pressing the "Create Pull Request" button, please provide the following:
  - [x] A description about what and why you are contributing, even if it's trivial.
  - [x] The issue number(s) or PR number(s) in the description if you are contributing in response to those.
  - [x] If applicable, unit tests.

### CODE REVIEW

File: Purchases/PurchasesExtensions/PurchaseOwnershipType+Extensions.swift [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+        let purchaseOwnershipTypeString = try container.decode(String.self)
+
+        switch purchaseOwnershipTypeString {
+        case "PURCHASED":
+            self = .purchased
+        case "FAMILY_SHARED":
+            self = .familyShared
+        default:
+            Logger.warn("received unknown ownershipType: \(purchaseOwnershipTypeString)")
+            self = .unknown
```
[the remark this answers is not in the document]
Comment (Juanpe) [reply Purchases/PurchasesExtensions/PurchaseOwnershipType+Extensions.swift:29]: I was working on this, and I realized that there is a special case, and I don't know exactly why. So I'd rather ask before assuming something wrongly 🙂

For both the `periodType` and the `Store` type, the default value is the same when the key is missing and when the value is unknown. But, this behaviour is different for `PurchaseOwnershipType`. If the key is missing, the default value is `purchased`, but if the value is unknown, the default value is `unknown`.

The question is, we could use a helper method to decode the value for the key and define the default value as well for both the `periodType` and the `Store` type and use the code that belongs to this PR for `PurchaseOwnershipType` or use the helper method for all the keys and simplify the logic for `PurchaseOwnershipType`.

WDYT? 🤔

[the remark this answers is not in the document]
Comment (aboedo) [reply Purchases/PurchasesExtensions/PurchaseOwnershipType+Extensions.swift:29]: We recently were talking about this here:
https://github.com/RevenueCat/purchases-ios/issues/775

The TL;DR: is that `ownershipType` is new, and before it existed, you could only just own something by purchasing it directly, so if the value is empty we assume that it was a direct purchase. 
But if we get a value that's different from the ones we know about, we might have to assume that there's a new way to own something (like if apple introduced a bulk purchase program for businesses, for example). 
We might wanna have an `@unknown default` for that case in case something new comes up and we forgot to add it here. 
As for the best way to handle both cases here, I don't have a strong preference, so I'm happy to go with whatever feels best to you.

[the remark this answers is not in the document]
Comment (Juanpe) [reply Purchases/PurchasesExtensions/PurchaseOwnershipType+Extensions.swift:29]: OK, understood it. I gonna submit the changes that includes my proposal that I think could fit well the current scenario 👍🏼

File: Purchases/CodableExtensions/EntitlementProductData+Extensions.swift [hunk 2]
```diff
@@ -28,27 +28,11 @@ extension EntitlementInfo.ProductData {
...
-            self.ownershipType = .purchased
-        }
-
-        if let store = try container.decodeIfPresent(Store.self, forKey: .store) {
-            self.store = store
-        } else {
-            Logger.warn("nil store found during decoding")
-            self.store = .unknownStore
-        }
+        self.periodType = container.decode(PeriodType.self, forKey: .periodType, defaultValue: .normal)
```
Comment (Juanpe) [inline Purchases/CodableExtensions/EntitlementProductData+Extensions.swift:31]: now, we are using the helper method to set the default value when the value is missing or decoding fails

File: Purchases/CodableExtensions/PurchaseOwnershipType+Extensions.swift [hunk 3]
```diff
@@ -17,15 +17,17 @@ extension PurchaseOwnershipType: Decodable {
...
 
         switch purchaseOwnershipTypeString {
         case "PURCHASED":
             self = .purchased
         case "FAMILY_SHARED":
             self = .familyShared
         default:
-            Logger.warn("received unknown ownershipType: \(purchaseOwnershipTypeString)")
+            Logger.error(Strings.codable.unexpectedValueError(for: PurchaseOwnershipType.self))
             self = .unknown
```
Comment (Juanpe) [inline Purchases/CodableExtensions/PurchaseOwnershipType+Extensions.swift:31]: This is a special case. Here we don't throw the error, we return the `unknown` case and log the error.

File: Purchases/FoundationExtensions/JSONDecoder+Extensions.swift [hunk 4]
```diff
@@ -33,15 +48,45 @@ extension JSONDecoder {
...
+    ///   - type: The type of value to decode.
+    ///   - key: The key that the decoded value is associated with.
+    ///   - defaultValue: The default value returned if decoding fails.
+    /// - Returns: A value of the requested type, or the given default value
+    /// if decoding fails.
+    func decode<T: Decodable>(_ type: T.Type, forKey key: Self.Key, defaultValue: T) -> T {
+        do {
+            return try decode(type, forKey: key)
+        } catch {
+            Logger.error(error.localizedDescription)
```
Comment (Juanpe) [inline Purchases/FoundationExtensions/JSONDecoder+Extensions.swift:87]: This error could be `value not found` or `value not expected`. Thus, we know why the decoding failed and we don't throw the error, so we return the given default value.

**Comment by Juanpe** [comment]
Thanks for the feedback ❤️ I've pushed the changes. Let me know if look fine for you 🤙

File: Purchases/FoundationExtensions/JSONDecoder+Extensions.swift (hunk 4 above)
Comment (taquitos) [reply Purchases/FoundationExtensions/JSONDecoder+Extensions.swift:87]: Excellent!

File: Purchases/Public/EntitlementInfo.swift [hunk 5]
```diff
@@ -355,3 +296,27 @@ private extension EntitlementInfo {
...
 
 }
+
+/**
+ This extension contains some internal helper structs to decode the data received from the backend.
+ */
+extension EntitlementInfo {
+
+    struct EntitlementData: Decodable {
+        let expiresDate: Date?
```
Comment (taquitos) [inline Purchases/Public/EntitlementInfo.swift:306]: Style: new lines for struct definition after `{` and before `}`

File: Purchases/Public/EntitlementInfo.swift [hunk 6]
```diff
@@ -355,3 +296,27 @@ private extension EntitlementInfo {
...
+        let purchaseDate: Date?
+        let productIdentifier: String
+    }
+
+    struct ProductData: Decodable {
+        let periodType: PeriodType
```
Comment (taquitos) [inline Purchases/Public/EntitlementInfo.swift:312]: same here

**Review by taquitos — CHANGES_REQUESTED** [review]
This looks great, one thing to do is to keep the folders/files alphabetized in the project 😄

File: Purchases/Public/EntitlementInfo.swift (hunk 6 above)
Comment (Juanpe) [reply Purchases/Public/EntitlementInfo.swift:326]: 👍🏼

File: Purchases/Public/EntitlementInfo.swift (hunk 5 above)
Comment (Juanpe) [reply Purchases/Public/EntitlementInfo.swift:318]: 👍🏼

**Comment by Juanpe** [comment]
> This looks great, one thing to do is to keep the folders/files alphabetized in the project 😄

Thanks! yep, you're right. I've done the changes 🙂

**Comment by taquitos** [comment]
> > This looks great, one thing to do is to keep the folders/files alphabetized in the project 😄
> 
> Thanks! yep, you're right. I've done the changes 🙂

The contents of the `CodableExtensions` folder still needs to be re-org'd 😄

**Comment by Juanpe** [comment]
> > > This looks great, one thing to do is to keep the folders/files alphabetized in the project 😄
> 
> > 
> 
> > Thanks! yep, you're right. I've done the changes 🙂
> 
> 
> 
> The contents of the `CodableExtensions` folder still needs to be re-org'd 😄 

Mmm... yep? What exactly do you mean? 🤔

This is the current order:
image

Is it not correct? Maybe I didn't understand you well 😅

**Comment by taquitos** [comment]
> > The contents of the `CodableExtensions` folder still needs to be re-org'd 😄
> 
> Mmm... yep? What exactly do you mean? 🤔
> 
> This is the current order:
> image
> 
> Is it not correct? Maybe I didn't understand you well 😅

These ones:
Screen Shot 2021-08-31 at 11 25 51 AM

**Comment by Juanpe** [comment]
> > > The contents of the `CodableExtensions` folder still needs to be re-org'd 😄
> 
> > 
> 
> > Mmm... yep? What exactly do you mean? 🤔
> 
> > 
> 
> > This is the current order:
> 
> > image
> 
> > 
> 
> > Is it not correct? Maybe I didn't understand you well 😅
> 
> 
> 
> These ones:
> 
> Screen Shot 2021-08-31 at 11 25 51 AM
> 
> 

Oohh shit! 🤦🏻‍♂️🤦🏻‍♂️

**Comment by taquitos** [comment]
LOL 😆

**Comment by Juanpe** [comment]
Well, I did some changes. Now, I think it's ok. Lets me know and forgive my foolishness 🤦🏼‍♂️ 😅

**Review by taquitos — APPROVED** [review]
🚀 🐐

**Review by aboedo — APPROVED** [review]

**Comment by aboedo** [comment]
Thanks for another fantastic PR @Juanpe!!
